### PR TITLE
Part 7 of Async Compaction : Ensure Compaction Operation compacts the data file as defined in the workload

### PR DESCRIPTION
**Pull request by bvaradar**
Part 7 of Async Compaction : Ensure Compaction Operation compacts the data file as defined in the workload

### CODE REVIEW

**Comment by bvaradar** [comment]
@n3nash @vinothchandar : Will be fixing the failing test (passes locally). But, ready otherwise for review

File: hoodie-client/src/main/java/com/uber/hoodie/io/HoodieMergeHandle.java [hunk 1]
```diff
@@ -71,30 +72,33 @@ public HoodieMergeHandle(HoodieWriteConfig config, String commitTime, HoodieTabl
...
-    init(fileId, init(fileId, recordItr));
+    String partitionPath = init(fileId, recordItr);
+    init(fileId, partitionPath,
+        fileSystemView.getLatestDataFiles(partitionPath)
+            .filter(dataFile -> dataFile.getFileId().equals(fileId)).findFirst());
   }
 
   public HoodieMergeHandle(HoodieWriteConfig config, String commitTime, HoodieTable<T> hoodieTable,
-      Map<String, HoodieRecord<T>> keyToNewRecords, String fileId) {
+      Map<String, HoodieRecord<T>> keyToNewRecords, String fileId, Optional<HoodieDataFile> oldDataFileToBeMerged) {
```
Comment (n3nash) [inline hoodie-client/src/main/java/com/uber/hoodie/io/HoodieMergeHandle.java]: We should avoid passing an extra HoodieDataFile here.

File: hoodie-client/src/main/java/com/uber/hoodie/io/compact/HoodieRealtimeTableCompactor.java [hunk 2]
```diff
@@ -122,8 +123,13 @@
 
     // Compacting is very similar to applying updates to existing file
     HoodieCopyOnWriteTable table = new HoodieCopyOnWriteTable(config, metaClient);
+
+    Optional<HoodieDataFile> oldDataFileOpt = table.getROFileSystemView()
```
Comment (n3nash) [inline hoodie-client/src/main/java/com/uber/hoodie/io/compact/HoodieRealtimeTableCompactor.java]: If required, you can perform this operation inside MergeHandle and remove passing it down.

**Comment by n3nash** [comment]
A refactor may be needed for the datafile argument, rest LGTM.

File: hoodie-client/src/main/java/com/uber/hoodie/io/HoodieMergeHandle.java [hunk 3]
```diff
@@ -71,30 +72,33 @@ public HoodieMergeHandle(HoodieWriteConfig config, String commitTime, HoodieTabl
...
     init(fileId, keyToNewRecords.get(keyToNewRecords.keySet().stream().findFirst().get())
-        .getPartitionPath());
+        .getPartitionPath(), oldDataFileToBeMerged);
   }
 
   /**
    * Extract old file path, initialize StorageWriter and WriteStatus
    */
-  private void init(String fileId, String partitionPath) {
+  private void init(String fileId, String partitionPath, Optional<HoodieDataFile> oldDataFileToBeMerged) {
```
Comment (vinothchandar) [inline hoodie-client/src/main/java/com/uber/hoodie/io/HoodieMergeHandle.java]: rename: dataFileToBeMerged

File: hoodie-client/src/main/java/com/uber/hoodie/io/compact/HoodieRealtimeTableCompactor.java (hunk 2 above)
Comment (vinothchandar) [reply hoodie-client/src/main/java/com/uber/hoodie/io/compact/HoodieRealtimeTableCompactor.java]: I asked Balaji to keep it like this. I actually think we should move our IOHandles to be dumb and just to an operation based on list of files passed in..  (similar to generating fileID outside CreateHandle)

File: hoodie-client/src/main/java/com/uber/hoodie/table/HoodieCopyOnWriteTable.java [hunk 4]
```diff
@@ -171,26 +171,26 @@ public HoodieCompactionWorkload scheduleCompaction(JavaSparkContext jsc, String
...
-    HoodieMergeHandle upsertHandle = getUpdateHandle(commitTime, fileLoc, recordItr);
-    return handleUpdateInternal(upsertHandle, commitTime, fileLoc);
+    HoodieMergeHandle upsertHandle = getUpdateHandle(commitTime, fileId, recordItr);
+    return handleUpdateInternal(upsertHandle, commitTime, fileId);
   }
 
-  public Iterator<List<WriteStatus>> handleUpdate(String commitTime, String fileLoc,
-      Map<String, HoodieRecord<T>> keyToNewRecords) throws IOException {
+  public Iterator<List<WriteStatus>> handleUpdate(String commitTime, String fileId,
+      Map<String, HoodieRecord<T>> keyToNewRecords, Optional<HoodieDataFile> oldDataFileOpt) throws IOException {
```
Comment (vinothchandar) [inline hoodie-client/src/main/java/com/uber/hoodie/table/HoodieCopyOnWriteTable.java]: please remove "old" from naming of all variables.. its pretty confusing to read. :)

Comment (bvaradar) [reply hoodie-client/src/main/java/com/uber/hoodie/table/HoodieCopyOnWriteTable.java]: Renamed

Comment (bvaradar) [reply hoodie-client/src/main/java/com/uber/hoodie/table/HoodieCopyOnWriteTable.java]: Renamed

File: hoodie-client/src/main/java/com/uber/hoodie/io/HoodieMergeHandle.java (hunk 3 above)
Comment (bvaradar) [reply hoodie-client/src/main/java/com/uber/hoodie/io/HoodieMergeHandle.java]: renamed

File: hoodie-client/src/main/java/com/uber/hoodie/io/HoodieMergeHandle.java (hunk 1 above)
Comment (bvaradar) [reply hoodie-client/src/main/java/com/uber/hoodie/io/HoodieMergeHandle.java]: this is in the same spirit as Vinoth explained

File: hoodie-client/src/main/java/com/uber/hoodie/io/compact/HoodieRealtimeTableCompactor.java (hunk 2 above)
Comment (n3nash) [reply hoodie-client/src/main/java/com/uber/hoodie/io/compact/HoodieRealtimeTableCompactor.java]: okay, we might need to perform a clean of this code then soon

**Comment by bvaradar** [comment]
Committed as part of Consolidated diff

**Pull request closed without merging**